Add `run_command_async`

justindbaur opened the pull request:
## 🎟️ Tracking

<!-- Paste the link to the Jira or GitHub issue or otherwise describe / point to where this change is coming from. -->

## 📔 Objective

<!-- Describe what the purpose of this PR is, for example what bug you're fixing or new feature you're adding. -->

- Bumps .NET to 8 (latest LTS)
- Uses `[LibraryImport]` for auto-marshalling of function
- Adds `run_command_async` method in `bitwarden-c` that allows for a non-blocking call that can notify the caller when the action is completed.
- `Task`-ifies one of the C# SDK calls into the SDK. 
  - This only updates one method to be async for now but all of them can be changed
  - This is a breaking change so would be a good idea before a `1.0.0` release.


## 📸 Screenshots

<!-- Required for any UI changes; delete if not applicable. Use fixed width images for better display. -->

## ⏰ Reminders before review

- Contributor guidelines followed
- All formatters and local linters executed and passed
- Written new unit and / or integration tests where applicable
- Protected functional changes with optionality (feature flags)
- Used internationalization (i18n) for all UI strings
- CI builds passed
- Communicated to DevOps any deployment requirements
- Updated any necessary documentation (Confluence, contributing docs) or informed the documentation
  team

## 🦮 Reviewer guidelines

<!-- Suggested interactions but feel free to use (or not) as you desire! -->

- 👍 (`:+1:`) or similar for great changes
- 📝 (`:memo:`) or ℹ️ (`:information_source:`) for notes or general info
- ❓ (`:question:`) for questions
- 🤔 (`:thinking:`) or 💭 (`:thought_balloon:`) for more open inquiry that's not quite a confirmed
  issue and could potentially benefit from discussion
- 🎨 (`:art:`) for suggestions / improvements
- ❌ (`:x:`) or ⚠️ (`:warning:`) for more significant problems or concerns needing attention
- 🌱 (`:seedling:`) or ♻️ (`:recycle:`) for future improvements or indications of technical debt
- ⛏ (`:pick:`) for minor or nitpick changes

dani-garcia commented on languages/csharp/Bitwarden.Sdk/Bitwarden.Sdk.csproj:
```diff
@@ -1,11 +1,11 @@
 <Project Sdk="Microsoft.NET.Sdk">
 
   <PropertyGroup>
-    <TargetFramework>net6.0</TargetFramework>
```
Question: Does increasing this affect people using the SDK in older .NET versions?

justindbaur replied:
Yes :/ .NET 6 does go out of support in November though. My current thinking is that I will try and convert more of the commands to be Async and then maybe we do a `1.0.0` release after .NET 6 goes out of support with the breaking changes of converting them.

Hinton replied:
Do we need to upgrade target framework? We've had requests to use NET Standard 2.0 to support .NET Framework.

justindbaur replied:
`[LibraryImport]` requires .NET 7+, in my opinion supporting .NET Framework would not be worth it. C# may soon be getting built in `Result` and `Option` types that we could take advantage of here and getting locked into .NET Framework support would either mean a of conditional compilation or just holding back the .NET targets.

dani-garcia commented on languages/csharp/Bitwarden.Sdk/BitwardenLibrary.cs:
```diff
@@ -2,20 +2,54 @@
+                }
+                finally
+                {
+                    Marshal.FreeCoTaskMem(resultPointer);
```
Is this freeing the Rust created result string? I don't think that's allowed according to the Rust docs, we'd need to free it from the Rust side:

https://doc.rust-lang.org/std/ffi/struct.CString.html#method.into_raw

Instead we should be freeing it from Rust after the callback was executed

justindbaur replied:
Good call, moved the freeing to rust.

dani-garcia reviewed (commented):
Overall LGTM, just some small details.

Hinton commented on languages/csharp/Bitwarden.Sdk/Bitwarden.Sdk.csproj:
```diff
@@ -1,11 +1,11 @@
     <Nullable>enable</Nullable>
     <RootNamespace>Bitwarden.Sdk</RootNamespace>
-
+    <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
```
Is anything using unsafe on the c# side?

justindbaur replied:
Yes, `[LibraryImport] does, it generates code like this that has unsafe blocks:

```csharp
private static partial void run_command_async(string json, global::Bitwarden.Sdk.BitwardenSafeHandle handle, global::Bitwarden.Sdk.BitwardenLibrary.OnCompleteCallback on_completed_callback)
{
    byte* __json_native = default;
    nint __handle_native = default;
    System.IntPtr __on_completed_callback_native = default;
    // Setup - Perform required setup.
    global::System.Runtime.InteropServices.Marshalling.SafeHandleMarshaller<global::Bitwarden.Sdk.BitwardenSafeHandle>.ManagedToUnmanagedIn __handle_native__marshaller = new();
    scoped global::System.Runtime.InteropServices.Marshalling.Utf8StringMarshaller.ManagedToUnmanagedIn __json_native__marshaller = new();
    try
    {
        // Marshal - Convert managed data to native data.
        __on_completed_callback_native = on_completed_callback != null ? global::System.Runtime.InteropServices.Marshal.GetFunctionPointerForDelegate(on_completed_callback) : default;
        __handle_native__marshaller.FromManaged(handle);
        __json_native__marshaller.FromManaged(json, stackalloc byte[global::System.Runtime.InteropServices.Marshalling.Utf8StringMarshaller.ManagedToUnmanagedIn.BufferSize]);
        {
            // PinnedMarshal - Convert managed data to native data that requires the managed data to be pinned.
            __handle_native = __handle_native__marshaller.ToUnmanaged();
            __json_native = __json_native__marshaller.ToUnmanaged();
            __PInvoke(__json_native, __handle_native, __on_completed_callback_native);
        }

        // NotifyForSuccessfulInvoke - Keep alive any managed objects that need to stay alive across the call.
        global::System.GC.KeepAlive(on_completed_callback);
    }
    finally
    {
        // CleanupCallerAllocated - Perform cleanup of caller allocated resources.
        __handle_native__marshaller.Free();
        __json_native__marshaller.Free();
    }

    // Local P/Invoke
    [global::System.Runtime.InteropServices.DllImportAttribute("bitwarden_c", EntryPoint = "run_command_async", ExactSpelling = true)]
    static extern unsafe void __PInvoke(byte* __json_native, nint __handle_native, System.IntPtr __on_completed_callback_native);
}
```

Hinton commented on languages/csharp/Bitwarden.Sdk/BitwardenLibrary.cs:
```diff
@@ -2,20 +2,47 @@
+            tcs.SetException(ex);
+        }
+
+        // TODO: Register cancellation on token
```
todo?

justindbaur replied:
It's still something I want to try out but I think it can be a future enhancement. 

It would require boxing the pointer to the returned `JoinHandle` on the rust side. Sending it back to C# to that we can register a cancellation action and then we likely have to add two new methods on the rust side. Probably `abort_and_drop_handle` and `drop_handle`. Do you have any feelings about that?

justindbaur replied:
I ended up doing this, let me know what you think about the implementation.

justindbaur commented on crates/bitwarden-core/Cargo.toml:
```diff
@@ -54,6 +54,9 @@ validator = { version = "0.18.1", features = ["derive"] }
 zeroize = { version = ">=1.7.0, <2.0", features = ["derive", "aarch64"] }
 zxcvbn = { version = ">=3.0.1, <4.0", optional = true }
 
+[target.'cfg(debug_assertions)'.dependencies]
+tokio = { version = "1.36.0", features = ["rt", "macros", "time"] }
```
I barely know what I did here, I need the time feature but only in debug, is this the right way to go?

Hinton replied:
It is, but I can't say I'm a fan of different behaviors on debug and release builds.

justindbaur replied:
This feels like the exact thing a debug build is for, it helps you write tests to make sure you've built the language wrapper correctly. That said I was able to move it into the `bitwarden-json` crate instead of `core`.

justindbaur commented:
I added support for `CancellationToken` which will cancel an ongoing call to rust. To ensure I did this correctly I added some debug only commands. I also converted all the C# methods to be `Task` returning and to optionally take a `CancellationToken`.

Hinton commented on crates/bitwarden-c/src/c.rs:
```diff
@@ -28,6 +33,48 @@ pub extern "C" fn run_command(c_str_ptr: *const c_char, client_ptr: *const CClie
+        }
+    });
+
+    // We only want to box the join handle the caller
+    // has said that they may want to cancel, essentially
+    // promising to us that they will take care of the
+    // returned pointer.
```
This seems fairly compact, `cargo fmt` generally uses 100 as the rule for line breaks.

```suggestion
    // We only want to box the join handle the caller has said that they may want to cancel,
    // essentially promising to us that they will take care of the returned pointer.
```

Hinton commented:
I added SM as reviewers, we should try and get this in before the 1.0 release of the C# sdk.

justindbaur commented on crates/bitwarden-core/src/platform/client_platform.rs:
```diff
@@ -24,6 +24,28 @@ impl<'a> ClientPlatform<'a> {
+        Err(Error::Internal(std::borrow::Cow::Borrowed(
+            "This is an error.",
+        )))
+    }
```
I moved them to `bitwarden-json`, if I moved it into `bitwarden-c` I would have to double parse the input string to check for special commands. Do you think that is alright?

Hinton replied:
Why not just expose two regular functions for it?

justindbaur replied:
The cancellation test is mostly testing the very code written in `run_command_async` so we'd have to expose a method with very similar internals to that and repeat ourselves. The error test is testing that the C# wrapper properly handles how errors are given back to it through the bitwarden-json crate.

Hinton commented on crates/bitwarden-json/src/command.rs:
```diff
@@ -188,3 +190,11 @@ pub enum GeneratorsCommand {
     GeneratePassword(PasswordGeneratorRequest),
 }
+
+#[cfg(debug_assertions)]
+#[derive(Serialize, Deserialize, JsonSchema, Debug)]
+#[serde(rename_all = "camelCase", deny_unknown_fields)]
+pub enum DebugCommand {
+    CancellationTest { duration_millis: u64 },
+    ErrorTest {},
+}
```
An alternative way. of doing this would be to add one or two separate commands to bitwarden-c.

Hinton reviewed (commented):
I think this seems fine from my perspective. I'll let SM do the final review and cover the csharp files.

coltonhurst reviewed (approved):
This looks good to me, thanks @justindbaur!

Have you tested that state files still work on the client creation and during token renewal?
